sql/async_udfs.mdx

```diff
@@ -0,0 +1,115 @@
+---
+title: Asynchronous User-Defined Functions
+description: 'Run Async UDFs'
+---
+
+Arroyo's UDFs can be asynchronous.
+This is useful for UDFs that need to make network requests or perform other asynchronous operations.
+To define an asynchronous UDF, use the `async` keyword in the function definition.
+
+## Options
+
+Async UDFs have a few configuration options that can be defined in a special comment at the top of the UDF,
+similar to dependencies, under a `[udfs]` header.
+
+| Option | Example | Description |
+| --- | --- | --- |
+| `async_results_ordered` | `true` | Whether to return results in the order they were requested (ordered) or as they are available (unordered). Defaults to true (ordered). |
+| `async_max_concurrency` | `10` | The maximum number of concurrent executions of this UDF. Defaults to 10. |
+| `async_timeout_seconds` | `10` | The maximum number of seconds to wait for a result before timing out. Arroyo currently drops the record if the UDF times out or panics. Defaults to 10. |
+
+
+## Context
+
+Async UDFs can optionally define a `Context` struct that will be passed to each execution of the UDF.
+This can be used to store resources that should be shared across executions of the UDF.
+
+The `Context` struct must implement a pre-defined trait called `UdfContext`.
+The trait is defined as follows:
+
+```rust
+#[async_trait]
+pub trait UdfContext: Sync {
+    async fn init(&self) {}
+    async fn close(&self) {}
+}
+```
+
+In addition to the above trait, the `Context` struct must also define a `new` function (with no arguments).
+The `new` and `init` functions will be called once when the job starts,
+and the `close` function will be called once when the job ends.
+
+If a `Context` is defined, an `Arc` reference to it will be passed as the first argument to the UDF.
+
+## Example
+
+This example defines an async UDF that's used to fetch a user's name from a database.
+
+```rust
+/*
+[dependencies]
+tokio = { version = "1", features = ["full"] }
+tokio-postgres = "0.7"
+async-trait = "0.1.68"
+arroyo-types = { path = "../../../arroyo-types" }
+
+[udfs]
+async_results_ordered = true
+async_max_concurrency = 10
+async_timeout_seconds = 5
+*/
+
+use tokio_postgres::{NoTls, Error, Client};
+use async_trait::async_trait;
+use arroyo_types::UdfContext;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+pub struct Context {
+    client: Mutex<Option<Client>>,
```

Comment:
Putting this on a Mutex means that only one call can have access to the client at any given time, which isn't great. Could this be done with a RWLock instead?

Reply:
Good call, done.